File: src/106_Shared_Mutex/reader_writer_pattern.cpp
/*
 * Reader-Writer Pattern Demo using std::shared_mutex
 * Compile: g++ -pthread --std=c++20 reader_writer_pattern.cpp -o app
 * 
 * Demonstrates:
 * - Multiple concurrent readers with shared_lock
 * - Exclusive writer access with unique_lock
 * - Thread-safe cache implementation
 */

#include <iostream>
#include <shared_mutex>
#include <thread>
#include <vector>
#include <string>
#include <mutex>
#include <unordered_map>

/**
 * Thread-safe cache using the Reader-Writer pattern
 * 
 * Uses std::shared_mutex to allow:
 * - Multiple simultaneous readers (shared access)
 * - Exclusive writer access (blocks all readers and other writers)
 */
class ThreadSafeCache {
private:
    // Mutable allows locking in const methods (read operations)
    mutable std::shared_mutex mutex_;
    
    // Underlying data structure protected by mutex
    std::unordered_map<std::string, std::string> cache_;

public:
    /**
     * Read operation - allows multiple concurrent readers
     * 
     * Uses shared_lock: multiple threads can hold this lock simultaneously
     * as long as no writer holds unique_lock
     * 
     * @param key The key to look up
     * @return The value if found, otherwise "Not found"
     */
    std::string read(const std::string& key) const {
        // Shared lock - multiple readers can acquire this simultaneously
        std::shared_lock<std::shared_mutex> lock(mutex_);
        
        auto it = cache_.find(key);
        return (it != cache_.end()) ? it->second : "Not found";
        
        // Lock automatically released when 'lock' goes out of scope (RAII)
    }

    /**
     * Write operation - exclusive access required
     * 
     * Uses unique_lock: only one thread can hold this lock
     * Blocks all readers and other writers until complete
     * 
     * @param key The key to insert/update
     * @param value The value to store
     */
    void write(const std::string& key, const std::string& value) {
        // Unique lock - exclusive access, blocks all other threads
        std::unique_lock<std::shared_mutex> lock(mutex_);
        
        cache_[key] = value;
        
        // Lock automatically released when 'lock' goes out of scope (RAII)
    }

    /**
     * Size query - read operation with shared access
     * 
     * @return Current number of entries in cache
     */
    size_t size() const {
        // Shared lock - can be called concurrently with other reads
        std::shared_lock<std::shared_mutex> lock(mutex_);
        return cache_.size();
    }
};

// Separate mutex to protect std::cout from interleaved output
// (std::cout is not thread-safe for concurrent writes)
std::mutex cout_mutex;

int main() {
    ThreadSafeCache cache;
    
    /**
     * Writer thread - populates cache with 5 key-value pairs
     * Writes every 100ms
     */
    std::thread writer([&cache]() {
        for (int i = 0; i < 5; ++i) {
            // Exclusive write - blocks all readers during this operation
            cache.write("key" + std::to_string(i), "value" + std::to_string(i));
            
            // Protect console output to prevent garbled text
            {
                std::lock_guard<std::mutex> lock(cout_mutex);
                std::cout << "Writer " << i << ": " << i << "\n";
            } // cout_mutex released here
            
            // Sleep to simulate real work and allow readers to interleave
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    });
    
    /**
     * Reader threads - 3 concurrent readers
     * Each attempts to read keys 0-4 (cycling), every 50ms
     * 
     * Note: Readers run faster (50ms) than writer (100ms),
     * so early reads may find "Not found" for keys not yet written
     */
    std::vector<std::thread> readers;
    for (int i = 0; i < 3; ++i) {
        readers.emplace_back([&cache, i]() {
            for (int j = 0; j < 10; ++j) {
                // Shared read - can run concurrently with other reads
                // but will block if writer holds unique_lock
                std::string value = cache.read("key" + std::to_string(j % 5));
                
                // Protect console output
                {
                    std::lock_guard<std::mutex> lock(cout_mutex);
                    std::cout << "Reader " << i << ": " << value << "\n";
                } // cout_mutex released here
                
                // Sleep shorter than writer, demonstrating race conditions
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }
        });
    }
    
    // Wait for writer to complete
    writer.join();
    
    // Wait for all readers to complete
    for (auto& reader : readers) {
        reader.join();
    }
    
    return 0;
}

File: src/106_Shared_Mutex/reader_writer_pattern.rs
/*
## Key Rust Equivalents

**C++ → Rust Mappings:**
- `std::shared_mutex` → `RwLock` (reader-writer lock)
- `std::shared_lock` → `RwLock::read()` (shared read access)
- `std::unique_lock` → `RwLock::write()` (exclusive write access)
- `std::mutex` → `Mutex` (for stdout protection)
- `std::thread` → `std::thread`
- Raw pointers/references → `Arc` (Atomic Reference Counting for shared ownership)

## Important Rust Differences

1. **Ownership with Arc**: Rust requires explicit shared ownership using `Arc` (like `shared_ptr` in C++), since multiple threads need access to the same cache

2. **No manual locking syntax**: Rust's RAII is even more explicit - the lock guard variable (`cache`) must exist to access the data

3. **Poisoning**: Rust locks can be "poisoned" if a thread panics while holding a lock, which is why we use `.unwrap()` (or could handle errors explicitly)

4. **Move semantics**: Variables are moved into closures with `move`, and we clone `Arc` pointers with `Arc::clone()`

5. **Type safety**: Rust's type system ensures thread safety at compile time - you literally cannot compile code that has data races

The behavior and output will be identical to the C++ version, including the "Not found" messages due to readers racing ahead of the writer!
*/

/*
 * Reader-Writer Pattern Demo using RwLock
 * Compile: rustc reader_writer_pattern.rs
 * Or with Cargo: cargo run
 * 
 * Demonstrates:
 * - Multiple concurrent readers with read()
 * - Exclusive writer access with write()
 * - Thread-safe cache implementation using Arc and RwLock
 */

use std::collections::HashMap;
use std::sync::{Arc, RwLock, Mutex};
use std::thread;
use std::time::Duration;

/**
 * Thread-safe cache using the Reader-Writer pattern
 * 
 * Uses RwLock to allow:
 * - Multiple simultaneous readers (shared access)
 * - Exclusive writer access (blocks all readers and other writers)
 * 
 * Wrapped in Arc for shared ownership across threads
 */
struct ThreadSafeCache {
    // RwLock allows multiple readers or one writer
    cache: RwLock<HashMap<String, String>>,
}

impl ThreadSafeCache {
    /**
     * Create a new empty cache
     */
    fn new() -> Self {
        ThreadSafeCache {
            cache: RwLock::new(HashMap::new()),
        }
    }

    /**
     * Read operation - allows multiple concurrent readers
     * 
     * Uses read(): multiple threads can hold read locks simultaneously
     * as long as no writer holds a write lock
     * 
     * @param key The key to look up
     * @return The value if found, otherwise "Not found"
     */
    fn read(&self, key: &str) -> String {
        // Acquire read lock - multiple readers can hold this simultaneously
        // .unwrap() panics if lock is poisoned (a thread panicked while holding lock)
        let cache = self.cache.read().unwrap();
        
        // Look up the key and return cloned value or "Not found"
        cache.get(key)
            .map(|v| v.clone())
            .unwrap_or_else(|| "Not found".to_string())
        
        // Read lock automatically released when 'cache' goes out of scope (RAII)
    }

    /**
     * Write operation - exclusive access required
     * 
     * Uses write(): only one thread can hold this lock
     * Blocks all readers and other writers until complete
     * 
     * @param key The key to insert/update
     * @param value The value to store
     */
    fn write(&self, key: String, value: String) {
        // Acquire write lock - exclusive access, blocks all other threads
        // .unwrap() panics if lock is poisoned
        let mut cache = self.cache.write().unwrap();
        
        cache.insert(key, value);
        
        // Write lock automatically released when 'cache' goes out of scope (RAII)
    }

    /**
     * Size query - read operation with shared access
     * 
     * @return Current number of entries in cache
     */
    fn size(&self) -> usize {
        // Acquire read lock - can be called concurrently with other reads
        let cache = self.cache.read().unwrap();
        cache.len()
    }
}

fn main() {
    // Arc (Atomic Reference Counting) allows shared ownership across threads
    let cache = Arc::new(ThreadSafeCache::new());
    
    // Mutex to protect stdout from interleaved output
    let cout_mutex = Arc::new(Mutex::new(()));

    /**
     * Writer thread - populates cache with 5 key-value pairs
     * Writes every 100ms
     */
    let cache_writer = Arc::clone(&cache);
    let cout_writer = Arc::clone(&cout_mutex);
    let writer = thread::spawn(move || {
        for i in 0..5 {
            // Exclusive write - blocks all readers during this operation
            cache_writer.write(
                format!("key{}", i),
                format!("value{}", i)
            );
            
            // Protect console output to prevent garbled text
            {
                let _lock = cout_writer.lock().unwrap();
                println!("Writer {}: {}", i, i);
            } // cout_mutex released here
            
            // Sleep to simulate real work and allow readers to interleave
            thread::sleep(Duration::from_millis(100));
        }
    });

    /**
     * Reader threads - 3 concurrent readers
     * Each attempts to read keys 0-4 (cycling), every 50ms
     * 
     * Note: Readers run faster (50ms) than writer (100ms),
     * so early reads may find "Not found" for keys not yet written
     */
    let mut readers = vec![];
    for i in 0..3 {
        let cache_reader = Arc::clone(&cache);
        let cout_reader = Arc::clone(&cout_mutex);
        
        let reader = thread::spawn(move || {
            for j in 0..10 {
                // Shared read - can run concurrently with other reads
                // but will block if writer holds write lock
                let value = cache_reader.read(&format!("key{}", j % 5));
                
                // Protect console output
                {
                    let _lock = cout_reader.lock().unwrap();
                    println!("Reader {}: {}", i, value);
                } // cout_mutex released here
                
                // Sleep shorter than writer, demonstrating race conditions
                thread::sleep(Duration::from_millis(50));
            }
        });
        
        readers.push(reader);
    }

    // Wait for writer to complete
    writer.join().unwrap();
    
    // Wait for all readers to complete
    for reader in readers {
        reader.join().unwrap();
    }
}